nemo_rl/models/policy/megatron_policy_worker.py

```diff
@@ -1801,21 +1806,63 @@ def generate(
             max_batch_size=self.cfg["generation_batch_size"],
         )
 
-        # detokenize the prompts
-        # detokenized_prompts = [
-        # self.tokenizer.decode(prompt)
-        # for prompt in data.get("input_ids")
-        # ]
-        # apply chat template
-        out = run_mcore_engine(
-            engine=inference_engine,
-            # prompts = detokenized_prompts,
-            prompt_tokens_tensor=data["input_ids"],
-            prompt_lengths_tensor=data["input_lengths"],
-            tokens_to_generate=self.cfg["generation"]["max_new_tokens"]  # type: ignore
-            - data["input_ids"].size(1),
+        input_ids = data["input_ids"]
+        tokens_to_generate = self.cfg["generation"]["max_new_tokens"] - input_ids.size(
+            1
+        )
+
+        padding = torch.full(
+            (input_ids.shape[0], tokens_to_generate),
+            self.megatron_tokenizer.eod_id,
+            dtype=input_ids.dtype,
+            device=input_ids.device,
+        )
+        prompt_tokens_tensor = torch.cat([input_ids, padding], dim=1)
+        prompt_lengths_tensor = data["input_lengths"]
+
+        # Handle None values for top_k - convert to integer as required by Megatron
+        top_k_cfg = self.cfg["generation"]["top_k"]
+        top_k_val = 1 if greedy else (int(top_k_cfg) if top_k_cfg is not None else 0)
+
+        # Use temperature 0.0 for greedy, 1.0 otherwise
+        temperature = 0.0 if greedy else 1.0
+
+        top_p_cfg = self.cfg["generation"]["top_p"]
+        top_p_val = (
+            0.0 if greedy else (float(top_p_cfg) if top_p_cfg is not None else 0.0)
         )
-        # print(out)
+
+        sampling_params = SamplingParams(
+            temperature=temperature,
+            top_k=top_k_val,
+            top_p=top_p_val,
+            return_segments=False,
+            return_log_probs=True,
+            num_tokens_to_generate=tokens_to_generate,
+            top_n_logprobs=0,
+            return_prompt_top_n_logprobs=False,
+        )
+        requests = []
+        for p, prompt_len in zip(
+            prompt_tokens_tensor, prompt_lengths_tensor, strict=True
+        ):
+            tokenized_prompt = p[:prompt_len].cpu().numpy().tolist()
```

Comment:
Can p have length greater than `prompt_len`? Also, curious why does InferenceRequest require detokenized prompts? They don't provide token-in / token-out API?

Reply:
Yes, the input data is already padded to the max prompt length. Not sure why. So had to cut it down again.